File: OrchidCore/src/main/java/com/eden/orchid/api/resources/resource/ResourceTransformation.java
package com.eden.orchid.api.resources.resource;

import com.eden.orchid.utilities.OrchidUtils;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

/**
 * A Resource type that wraps another resource, optionally applying a transformation along the way.
 */
public class ResourceTransformation extends ResourceWrapper {

    private final List<Function<String, String>> contentTransformations;
    private final List<Function<InputStream, InputStream>> contentStreamTransformations;

    public ResourceTransformation(OrchidResource resource, Function<String, String>... contentTransformations) {
        this(resource, Arrays.asList(contentTransformations), null);
    }

    public ResourceTransformation(OrchidResource resource, List<Function<String, String>> contentTransformations, List<Function<InputStream, InputStream>> contentStreamTransformations) {
        super(resource);
        this.contentTransformations = (contentTransformations != null) ? contentTransformations : new ArrayList<>();
        this.contentStreamTransformations = (contentStreamTransformations != null) ? contentStreamTransformations : new ArrayList<>();
    }

    @Override
    public String getContent() {
        return OrchidUtils.transform(super.getContent(), contentTransformations);
    }

    @Override
    public InputStream getContentStream() {
        return OrchidUtils.transform(super.getContentStream(), contentStreamTransformations);
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/resources/ResourceTransformation.kt
package com.eden.orchid.api.resources

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.resources.resource.OrchidResource
import com.eden.orchid.api.resources.resource.ResourceWrapper
import java.io.InputStream

/**
 * A Resource type that wraps another resource, optionally applying a transformation along the way.
 */
open class ResourceTransformation

@JvmOverloads
constructor(
        resource: OrchidResource,
        protected var contentPreTransformations: MutableList<(String) -> String> = mutableListOf(),
        protected var contentPostTransformations: MutableList<(String) -> String> = mutableListOf(),
        protected var contentStreamTransformations: MutableList<(InputStream) -> InputStream> = mutableListOf()
) : ResourceWrapper(resource) {

    override fun getContent(): String {
        return transform(super.getContent(), contentPreTransformations)
    }

    override fun getContentStream(): InputStream {
        return transform(super.getContentStream(), contentStreamTransformations)
    }

    override fun compileContent(data: Any?): String? {
        return transform(super.compileContent(data), contentPostTransformations)
    }

    fun <T> transform(input: T, transformations: List<(T) -> T>): T {
        return if(transformations.isEmpty()) input else transformations.reduce { a, b -> { o: T -> b.invoke(a.invoke(o)) } } .invoke(input)
    }

}
